presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java

```diff
@@ -42,29 +41,45 @@ private WorkProcessorUtils() {}
     {
         return new AbstractIterator<T>()
         {
+            final Iterator<Optional<T>> yieldingIterator = yieldingIteratorFrom(processor);
+
             @Override
             protected T computeNext()
             {
                 while (true) {
-                    if (processor.process()) {
-                        if (processor.isFinished()) {
-                            return endOfData();
-                        }
+                    if (!yieldingIterator.hasNext()) {
+                        return endOfData();
+                    }
 
-                        return processor.getResult();
+                    Optional<T> elementOptional = yieldingIterator.next();
+                    if (elementOptional.isPresent()) {
+                        return elementOptional.get();
                     }
-                    else if (processor.isBlocked()) {
-                        try {
-                            processor.getBlockedFuture().get();
-                        }
-                        catch (InterruptedException e) {
-                            Thread.currentThread().interrupt();
-                            throw new RuntimeException(e);
-                        }
-                        catch (ExecutionException e) {
-                            throw new RuntimeException(e);
-                        }
+                }
+            }
+        };
+    }
+
+    static <T> Iterator<Optional<T>> yieldingIteratorFrom(WorkProcessor<T> processor)
+    {
+        return new AbstractIterator<Optional<T>>()
+        {
+            @Override
+            protected Optional<T> computeNext()
+            {
+                if (processor.process()) {
+                    if (processor.isFinished()) {
+                        return endOfData();
                     }
+
+                    return Optional.of(processor.getResult());
+                }
+                else if (processor.isBlocked()) {
+                    throw new IllegalStateException("Work is blocked");
```

Comment:
Why do you just throw here? It used to wait:
```
 -                        try {
 -                            processor.getBlockedFuture().get();
 -                        }
 -                        catch (InterruptedException e) {
 -                            Thread.currentThread().interrupt();
 -                            throw new RuntimeException(e);
 -                        }
 -                        catch (ExecutionException e) {
 -                            throw new RuntimeException(e);
 -                        }
```

Reply:
We discussed that offline. Hanging on a blocking future is a wrong thing to do, always. It leads to under-utilization of the processing power and, in extreme cases, may lead to starvation if e.g. all threads in a pool are waiting on a future which will be satisfied only when another task gets into the pool

Reply:
Having said that...
- it would be better to have this in a separate (first) commit, where argumentation can be attached to the commit message
- the exception could have a better message, like `"Cannot iterate over blocking WorkProcessor " + processor`